### ISSUE TITLE
Realm Dependent Partitioning Operations in Need of Tuning at Scale

### ISSUE THREAD

**Issue by lightsighter** [StanfordLegion/realm#19]
Some Realm dependent partitioning operations are exhibiting poor asymptotic behavior when run at scale. When weak scaling Pennant across a 16 nodes we see the following behavior for partition-by-field operations:

1 node: ~0ms per operation (72M zones)
2 nodes: 170 - 230 ms per op (144M zones)
4 nodes: 850 ms - 1.5s per op (288M zones)
8 nodes: 5s - 32s per op (576M zones)
16 nodes: 62 - 858s per op (1.15B zones)

This is clearly super-linear in the size of the data being partitioned which is growing with O(N) in the number of nodes. In practice I expect Realm to discover the structure of the resulting partition which in this case is "nearest-neighbors" for an "unstructured" mesh, so in this case I would actually expect the partitioning time to be nearly constant as each node should just iterate over its physical instance and find the colors of the resulting points in its instance and then send messages to the few (but bounded) nearest neighbor nodes that have adjacent index spaces and need to be notified of some points.

This matters for Pennant now, but will matter for FleCSI in the very near future when we go to run meshes at scale on Sierra.

Requesting comment from @magnatelee since he has also seen slow Realm dependent partitioning operations in another context.

**Comment by lightsighter** [comment]
Yes, still very relevant. All the dependent partitioning operations are still not tuned and are much slower than they could be. They just aren't on the critical path of many things right now, but that doesn't mean that they are optimized to the level that they ultimately should be.

**Comment by lightsighter** [comment]
The initialization of the circuit simulation with some large graphs would definitely be a good proxy application as it performs lots of dependent partitioning operations.

**Comment by elliottslaughter** [comment]
It depends on which version of Circuit you use. Some versions of Circuit were "optimized" by over-approximating the set of rectangles required. In the case of `circuit_sparse.rg`, for example, we just made a bounding box around everything that might be required. Even though that multiplies that total data moved by some small integer, it's still a win overall because the rest of this infrastructure is so inefficient (both the partitioning ops and more importantly copies).

Anyway, we have several versions of Circuit and some of them do appear to do it the naive way (e.g., `language/examples/circuit/circuit_dep_par*.rg`). However you will still want to double check exactly what you're getting.

**Comment by lightsighter** [comment]
To be clear, I was talking about the C++ version of circuit which will exercise plenty of dependent partitioning operators (almost all of them).
